src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -268,6 +272,34 @@ private void CheckHeaderSignatureWithCoinstakeKernel(ProvenBlockHeader header, U
             ConsensusErrors.BadBlockSignature.Throw();
         }
 
+        /// <summary>
+        /// Checks if coinstake is spent on another chain.
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        /// <param name="context">The POS rule context.</param>
+        /// <exception cref="ConsensusException">Throws utxo not found error.</exception>
+        private void CheckIfCoinstakeIsSpentOnAnotherChain(ProvenBlockHeader header, PosRuleContext context)
+        {
+            Transaction coinstake = header.Coinstake;
+            TxIn input = coinstake.Inputs[0];
+
+            int? rewindDataIndex = this.PosParent.RewindDataIndexStore.Get(input.PrevOut.Hash, (int)input.PrevOut.N);
+            if (!rewindDataIndex.HasValue)
+            {
+                context.ValidationContext.SetFlagAndThrow(ConsensusErrors.ReadTxPrevFailed, c => c.InsufficientHeaderInformation = true);
+            }
+
+            RewindData rewindData = this.PosParent.UtxoSet.GetRewindData(rewindDataIndex.Value).GetAwaiter().GetResult();
+            UnspentOutputs matchingUnspentUtxo =
+                rewindData.OutputsToRestore.FirstOrDefault(unspent => unspent.TransactionId == input.PrevOut.Hash);
```

Comment:
Just wondering - it is possible that there is more than one UTXO from the same transaction in `OutputsToRestore`. We should match on the index as well.

Reply:
agre